pkg/controller/elasticsearch/validation/claim_validation.go

```diff
@@ -0,0 +1,52 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package validation
+
+import (
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// noIllegalVolumeClaimDeletePolicyChange validates that the user is not changing VolumeClaimDeletePolicy on an existing
+// Elasticsearch cluster. This is implemented as a create validation and not as an update validation to allow usage from
+// within the controller as well where we don't have the context of the previous version of the spec anymore.
+// But we can infer a policy change by checking the ownerReferences on the PVCs in the knowledge that a Retain policy
+// must mean no owner reference on the PVCs and vice versa for the Remove* policies
+func noIllegalVolumeClaimDeletePolicyChange(k8sClient k8s.Client, es esv1.Elasticsearch) field.ErrorList {
+	var errs field.ErrorList
+	matchingLabels := label.NewLabelSelectorForElasticsearch(es)
+	var pvcs v1.PersistentVolumeClaimList
```

Comment:
We should inspect StatefulSets here, not PersistentVolumeClaims.
(PVCs _are_ mutable, but not sset spec claims)

Reply:
This also means we could hack around sset spec immutability using a similar mechanism as we use for pvc resize, and remove pvc ownerRefs. But it's maybe not worth it...